File: src/VB/YandexDiskSDK/YandexDiskSDK/YClient.vb
Imports YandexDiskSDK.utilitiez
Imports Newtonsoft.Json
Imports YandexDiskSDK.JSON
Imports Newtonsoft.Json.Linq

Public Class YClient
    Implements IClient


    Public Sub New(ByRef accessToken As String, Optional Settings As ConnectionSettings = Nothing)
        authToken = accessToken
        'Destination_Type = PathPattern
        ConnectionSetting = Settings
        If Settings Is Nothing Then
            m_proxy = Nothing
        Else
            m_proxy = Settings.Proxy
            m_CloseConnection = If(Settings.CloseConnection, True)
            m_TimeOut = If(Settings.TimeOut = Nothing, TimeSpan.FromMinutes(60), Settings.TimeOut)
        End If
        Net.ServicePointManager.Expect100Continue = True : Net.ServicePointManager.SecurityProtocol = Net.SecurityProtocolType.Tls Or Net.SecurityProtocolType.Tls11 Or Net.SecurityProtocolType.Tls12 Or Net.SecurityProtocolType.Ssl3
    End Sub


    Public ReadOnly Property Item(Path As String) As IItem Implements IClient.Item
        Get
            Return New ItemClient(Path)
        End Get
    End Property
    Public ReadOnly Property Sharing() As ISharing Implements IClient.Sharing
        Get
            Return New SharingClient()
        End Get
    End Property
    Public ReadOnly Property RecycleBin(Path As String) As IRecycleBin Implements IClient.RecycleBin
        Get
            Return New RecycleBinClient(Path)
        End Get
    End Property


    Public ReadOnly Property RootPath(PathPattern As DestinationType) As String Implements IClient.RootPath
        Get
            Select Case PathPattern
                Case DestinationType.app
                    Return "app:/"
                Case DestinationType.disk
                    Return "disk:/"
                Case DestinationType.trash
                    Return "trash:/"
            End Select
        End Get
    End Property

#Region "UserInfo"
    Public Async Function _UserInfo() As Task(Of JSON_UserInfo) Implements IClient.UserInfo
        Using localHttpClient As New HttpClient(New HCHandler)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(New pUri("")).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_UserInfo)(result, JSONhandler)
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region



#Region "ListLatestUploadedFiles"
    Public Async Function Get_LatestFilesUploaded(Optional Filter As FilterEnum? = Nothing, Optional Fields As Fields = Fields.nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As PreviewSizeEnum = PreviewSizeEnum.S_150, Optional Limit As Integer = 200) As Task(Of JSON_FilesList) Implements IClient.ListLatestUploadedFiles
        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("media_type", Filter.ToString)
        parameters.Add("preview_crop", PreviewCrop)
        parameters.Add("preview_size", stringValueOf(PreviewSize))
        parameters.Add("fields", Fields.ToString)
        parameters.Add("limit", Limit)

        Using localHttpClient As New HttpClient(New HCHandler)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(New pUri("resources/last-uploaded", parameters)).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.IsSuccessStatusCode Then
                    Return JsonConvert.DeserializeObject(Of JSON_FilesList)(result, JSONhandler)
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region


#Region "CopyingMovingDeletingUrluploadingOperationStatus"
    Public Async Function _CheckOperationStatus(OperationHref As String) As Task(Of JSON_CheckOperationStatus) Implements IClient.CheckOperationStatus
        Using localHttpClient As New HttpClient(New HCHandler)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(New Uri(OperationHref)).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_CheckOperationStatus)(result, JSONhandler)
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region


End Class

File: src/Csharp/YandexDiskSDK/YandexDiskSDK/YClient.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using YandexDiskSDK.JSON;
using static YandexDiskSDK.utilitiez;

namespace YandexDiskSDK
{
    public class YClient : IClient
    {
        public YClient(string accessToken, ConnectionSettings Settings = null)
        {
            Base.authToken = accessToken;
            Base.ConnectionSetting = Settings;
            if (Settings == null) { Base.m_proxy = null; }
            else
            {
                Base.m_proxy = Settings.Proxy;
                Base.m_CloseConnection = Settings.CloseConnection ?? true;
                Base.m_TimeOut = Settings.TimeOut ?? TimeSpan.FromMinutes(60);
            }
            ServicePointManager.Expect100Continue = true; ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls12 | SecurityProtocolType.Ssl3;
        }

        public IItem Item(string Path) => new ItemClient(Path);
        public ISharing Sharing() => new SharingClient();
        public IRecycleBin RecycleBin(string Path) => new RecycleBinClient(Path);


        public string RootPath(DestinationType PathPattern)
        {
            string pth = string.Empty;
            switch (PathPattern)
            {
                case DestinationType.app:
                    pth = "app:/";
                    break;
                case DestinationType.disk:
                    pth = "disk:/";
                    break;
                case DestinationType.trash:
                    pth = "trash:/";
                    break;
            }
            return pth;
        }

        #region UserInfo
        public async Task<JSON_UserInfo> UserInfo()
        {
            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                using (HttpResponseMessage response = await localHttpClient.GetAsync(new Base.pUri("")).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();

                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        return JsonConvert.DeserializeObject<JSON_UserInfo>(result, Base.JSONhandler);
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "ListLatestUploadedFiles"
        public async Task<JSON_FilesList> ListLatestUploadedFiles(FilterEnum? Filter = null, Fields Fields = Fields.nothing, bool PreviewCrop = true, PreviewSizeEnum PreviewSize = PreviewSizeEnum.S_150, int Limit = 200)
        {
            var parameters = new Dictionary<string, string>
            {
                { "media_type", Filter.ToString() },
                { "preview_crop", PreviewCrop.ToString() },
                { "preview_size", stringValueOf(PreviewSize) },
                { "fields", Fields.ToString() },
                { "limit", Limit.ToString() }
            };
            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                using (HttpResponseMessage response = await localHttpClient.GetAsync(new Base.pUri("resources/last-uploaded", parameters)).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();

                    if (response.IsSuccessStatusCode)
                    {
                        return JsonConvert.DeserializeObject<JSON_FilesList>(result, Base.JSONhandler);
                    }

                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }


                }
            }
        }
        #endregion

        #region "CopyingMovingDeletingUrluploadingOperationStatus"
        public async Task<JSON_CheckOperationStatus> CheckOperationStatus(string OperationHref)
        {
            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                using (HttpResponseMessage response = await localHttpClient.GetAsync(new Uri(OperationHref)).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();
                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        return JsonConvert.DeserializeObject<JSON_CheckOperationStatus>(result, Base.JSONhandler);
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion



    }
}
